File: admin_app/admin_flutter/lib/models/stock_entry.dart
class StockEntry {
  const StockEntry({
    required this.id,
    this.supplierId,
    required this.source,
    required this.status,
    this.invoiceKey,
    this.invoiceNumber,
    this.invoiceSeries,
    this.supplierName,
    this.supplierDocument,
    required this.totalAmount,
    this.notes,
    required this.items,
    this.createdAt,
    this.confirmedAt,
  });

  final int id;
  final int? supplierId;
  final String source;
  final String status;
  final String? invoiceKey;
  final String? invoiceNumber;
  final String? invoiceSeries;
  final String? supplierName;
  final String? supplierDocument;
  final double totalAmount;
  final String? notes;
  final List<StockEntryItem> items;
  final DateTime? createdAt;
  final DateTime? confirmedAt;

  factory StockEntry.fromJson(Map<String, dynamic> json) {
    return StockEntry(
      id: json['id'] as int,
      supplierId: json['supplier_id'] as int?,
      source: json['source'] as String? ?? 'manual',
      status: json['status'] as String? ?? 'confirmed',
      invoiceKey: json['invoice_key'] as String?,
      invoiceNumber: json['invoice_number'] as String?,
      invoiceSeries: json['invoice_series'] as String?,
      supplierName: json['supplier_name'] as String?,
      supplierDocument: json['supplier_document'] as String?,
      totalAmount: _toDouble(json['total_amount']),
      notes: json['notes'] as String?,
      items: (json['items'] as List<dynamic>? ?? [])
          .map((item) => StockEntryItem.fromJson(item as Map<String, dynamic>))
          .toList(),
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? ''),
      confirmedAt: DateTime.tryParse(json['confirmed_at']?.toString() ?? ''),
    );
  }
}

class StockEntryItem {
  const StockEntryItem({
    this.productId,
    required this.description,
    this.barcode,
    this.invoiceQuantity,
    this.invoiceUnit,
    this.packageConversionFactor,
    required this.quantity,
    this.receivedQuantity,
    required this.unit,
    required this.unitCost,
    required this.totalCost,
    this.ncm,
    this.cfop,
    this.ibsCbsCst,
    this.ibsCbsClassification,
    this.cbsRate,
    this.ibsStateRate,
    this.ibsCityRate,
    this.selectiveTaxCst,
    this.selectiveTaxClassification,
    this.selectiveTaxRate,
    this.batchNumber,
    this.expirationDate,
    this.checkStatus = 'accepted',
    this.checkNotes,
  });

  final int? productId;
  final String description;
  final String? barcode;
  final double? invoiceQuantity;
  final String? invoiceUnit;
  final double? packageConversionFactor;
  final double quantity;
  final double? receivedQuantity;
  final String unit;
  final double unitCost;
  final double totalCost;
  final String? ncm;
  final String? cfop;
  final String? ibsCbsCst;
  final String? ibsCbsClassification;
  final double? cbsRate;
  final double? ibsStateRate;
  final double? ibsCityRate;
  final String? selectiveTaxCst;
  final String? selectiveTaxClassification;
  final double? selectiveTaxRate;
  final String? batchNumber;
  final String? expirationDate;
  final String checkStatus;
  final String? checkNotes;

  factory StockEntryItem.fromJson(Map<String, dynamic> json) {
    return StockEntryItem(
      productId: json['product_id'] as int?,
      description: json['description'] as String,
      barcode: json['barcode'] as String?,
      invoiceQuantity: _toNullableDouble(json['invoice_quantity']),
      invoiceUnit: json['invoice_unit'] as String?,
      packageConversionFactor: _toNullableDouble(
        json['package_conversion_factor'],
      ),
      quantity: _toDouble(json['quantity']),
      receivedQuantity: _toNullableDouble(json['received_quantity']),
      unit: json['unit'] as String? ?? 'un',
      unitCost: _toDouble(json['unit_cost']),
      totalCost: _toDouble(json['total_cost']),
      ncm: json['ncm'] as String?,
      cfop: json['cfop'] as String?,
      ibsCbsCst: json['ibs_cbs_cst'] as String?,
      ibsCbsClassification: json['ibs_cbs_classification'] as String?,
      cbsRate: _toNullableDouble(json['cbs_rate']),
      ibsStateRate: _toNullableDouble(json['ibs_state_rate']),
      ibsCityRate: _toNullableDouble(json['ibs_city_rate']),
      selectiveTaxCst: json['selective_tax_cst'] as String?,
      selectiveTaxClassification: json['selective_tax_classification'] as String?,
      selectiveTaxRate: _toNullableDouble(json['selective_tax_rate']),
      batchNumber: json['batch_number'] as String?,
      expirationDate: json['expiration_date'] as String?,
      checkStatus: json['check_status'] as String? ?? 'accepted',
      checkNotes: json['check_notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'product_id': productId,
      'description': description,
      'barcode': barcode,
      'invoice_quantity': invoiceQuantity,
      'invoice_unit': invoiceUnit,
      'package_conversion_factor': packageConversionFactor,
      'quantity': quantity,
      'received_quantity': receivedQuantity,
      'unit': unit,
      'unit_cost': unitCost,
      'total_cost': totalCost,
      'ncm': ncm,
      'cfop': cfop,
      'ibs_cbs_cst': ibsCbsCst,
      'ibs_cbs_classification': ibsCbsClassification,
      'cbs_rate': cbsRate,
      'ibs_state_rate': ibsStateRate,
      'ibs_city_rate': ibsCityRate,
      'selective_tax_cst': selectiveTaxCst,
      'selective_tax_classification': selectiveTaxClassification,
      'selective_tax_rate': selectiveTaxRate,
      'batch_number': batchNumber,
      'expiration_date': expirationDate,
      'check_status': checkStatus,
      'check_notes': checkNotes,
    };
  }
}

class StockEntryPayload {
  const StockEntryPayload({
    this.supplierId,
    this.supplierName,
    this.supplierDocument,
    required this.source,
    this.invoiceKey,
    this.invoiceNumber,
    this.invoiceSeries,
    this.notes,
    required this.items,
  });

  final int? supplierId;
  final String? supplierName;
  final String? supplierDocument;
  final String source;
  final String? invoiceKey;
  final String? invoiceNumber;
  final String? invoiceSeries;
  final String? notes;
  final List<StockEntryItem> items;

  Map<String, dynamic> toJson() {
    return {
      'supplier_id': supplierId,
      'supplier_name': supplierName,
      'supplier_document': supplierDocument,
      'source': source,
      'invoice_key': invoiceKey,
      'invoice_number': invoiceNumber,
      'invoice_series': invoiceSeries,
      'notes': notes,
      'items': items.map((item) => item.toJson()).toList(),
    };
  }
}

class StockEntryMobileItemPayload {
  const StockEntryMobileItemPayload({
    this.productId,
    required this.description,
    this.barcode,
    required this.quantity,
    required this.unit,
    required this.unitCost,
    this.salePrice,
    this.ncm,
    this.cfop,
    this.ibsCbsCst,
    this.ibsCbsClassification,
    this.cbsRate,
    this.ibsStateRate,
    this.ibsCityRate,
    this.selectiveTaxCst,
    this.selectiveTaxClassification,
    this.selectiveTaxRate,
    this.batchNumber,
    this.expirationDate,
    this.checkNotes,
  });

  final int? productId;
  final String description;
  final String? barcode;
  final double quantity;
  final String unit;
  final double unitCost;
  final double? salePrice;
  final String? ncm;
  final String? cfop;
  final String? ibsCbsCst;
  final String? ibsCbsClassification;
  final double? cbsRate;
  final double? ibsStateRate;
  final double? ibsCityRate;
  final String? selectiveTaxCst;
  final String? selectiveTaxClassification;
  final double? selectiveTaxRate;
  final String? batchNumber;
  final String? expirationDate;
  final String? checkNotes;

  Map<String, dynamic> toJson() {
    return {
      'product_id': productId,
      'description': description,
      'barcode': barcode,
      'quantity': quantity,
      'unit': unit,
      'unit_cost': unitCost,
      'sale_price': salePrice,
      'ncm': ncm,
      'cfop': cfop,
      'ibs_cbs_cst': ibsCbsCst,
      'ibs_cbs_classification': ibsCbsClassification,
      'cbs_rate': cbsRate,
      'ibs_state_rate': ibsStateRate,
      'ibs_city_rate': ibsCityRate,
      'selective_tax_cst': selectiveTaxCst,
      'selective_tax_classification': selectiveTaxClassification,
      'selective_tax_rate': selectiveTaxRate,
      'batch_number': batchNumber,
      'expiration_date': expirationDate,
      'check_notes': checkNotes,
    };
  }
}

class NfeXmlPreview {
  const NfeXmlPreview({
    this.supplierId,
    this.supplierName,
    this.supplierDocument,
    this.invoiceKey,
    this.invoiceNumber,
    this.invoiceSeries,
    required this.items,
  });

  final int? supplierId;
  final String? supplierName;
  final String? supplierDocument;
  final String? invoiceKey;
  final String? invoiceNumber;
  final String? invoiceSeries;
  final List<NfeXmlPreviewItem> items;

  factory NfeXmlPreview.fromJson(Map<String, dynamic> json) {
    return NfeXmlPreview(
      supplierId: json['supplier_id'] as int?,
      supplierName: json['supplier_name'] as String?,
      supplierDocument: json['supplier_document'] as String?,
      invoiceKey: json['invoice_key'] as String?,
      invoiceNumber: json['invoice_number'] as String?,
      invoiceSeries: json['invoice_series'] as String?,
      items: (json['items'] as List<dynamic>? ?? [])
          .map(
            (item) => NfeXmlPreviewItem.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
    );
  }
}

class NfeXmlPreviewItem {
  const NfeXmlPreviewItem({
    this.productId,
    this.productName,
    required this.description,
    this.barcode,
    required this.quantity,
    required this.unit,
    required this.unitCost,
    required this.totalCost,
    this.ncm,
    this.cfop,
    this.ibsCbsCst,
    this.ibsCbsClassification,
    this.cbsRate,
    this.ibsStateRate,
    this.ibsCityRate,
    this.selectiveTaxCst,
    this.selectiveTaxClassification,
    this.selectiveTaxRate,
    this.batchNumber,
    this.expirationDate,
  });

  final int? productId;
  final String? productName;
  final String description;
  final String? barcode;
  final double quantity;
  final String unit;
  final double unitCost;
  final double totalCost;
  final String? ncm;
  final String? cfop;
  final String? ibsCbsCst;
  final String? ibsCbsClassification;
  final double? cbsRate;
  final double? ibsStateRate;
  final double? ibsCityRate;
  final String? selectiveTaxCst;
  final String? selectiveTaxClassification;
  final double? selectiveTaxRate;
  final String? batchNumber;
  final String? expirationDate;

  factory NfeXmlPreviewItem.fromJson(Map<String, dynamic> json) {
    return NfeXmlPreviewItem(
      productId: json['product_id'] as int?,
      productName: json['product_name'] as String?,
      description: json['description'] as String,
      barcode: json['barcode'] as String?,
      quantity: _toDouble(json['quantity']),
      unit: json['unit'] as String? ?? 'un',
      unitCost: _toDouble(json['unit_cost']),
      totalCost: _toDouble(json['total_cost']),
      ncm: json['ncm'] as String?,
      cfop: json['cfop'] as String?,
      ibsCbsCst: json['ibs_cbs_cst'] as String?,
      ibsCbsClassification: json['ibs_cbs_classification'] as String?,
      cbsRate: _toNullableDouble(json['cbs_rate']),
      ibsStateRate: _toNullableDouble(json['ibs_state_rate']),
      ibsCityRate: _toNullableDouble(json['ibs_city_rate']),
      selectiveTaxCst: json['selective_tax_cst'] as String?,
      selectiveTaxClassification: json['selective_tax_classification'] as String?,
      selectiveTaxRate: _toNullableDouble(json['selective_tax_rate']),
      batchNumber: json['batch_number'] as String?,
      expirationDate: json['expiration_date'] as String?,
    );
  }
}

double _toDouble(Object? value) {
  return _toNullableDouble(value) ?? 0;
}

double? _toNullableDouble(Object? value) {
  if (value == null) return null;
  if (value is num) return value.toDouble();
  return double.tryParse(value.toString()) ?? 0;
}

class XmlInboxSettings {
  const XmlInboxSettings({required this.emailAddress, required this.enabled});

  final String emailAddress;
  final bool enabled;

  factory XmlInboxSettings.fromJson(Map<String, dynamic> json) {
    return XmlInboxSettings(
      emailAddress: json['email_address']?.toString() ?? '',
      enabled: json['enabled'] as bool? ?? false,
    );
  }
}

class XmlInboxMessage {
  const XmlInboxMessage({
    required this.id,
    this.stockEntryId,
    required this.status,
    this.senderEmail,
    this.subject,
    this.attachmentName,
    this.supplierName,
    this.supplierDocument,
    this.recipientDocument,
    this.invoiceKey,
    this.invoiceNumber,
    this.rejectionReason,
    this.receivedAt,
  });

  final int id;
  final int? stockEntryId;
  final String status;
  final String? senderEmail;
  final String? subject;
  final String? attachmentName;
  final String? supplierName;
  final String? supplierDocument;
  final String? recipientDocument;
  final String? invoiceKey;
  final String? invoiceNumber;
  final String? rejectionReason;
  final DateTime? receivedAt;

  bool get imported => status == 'imported';
  bool get pendingReceipt => status == 'pending_receipt';

  factory XmlInboxMessage.fromJson(Map<String, dynamic> json) {
    return XmlInboxMessage(
      id: json['id'] as int,
      stockEntryId: json['stock_entry_id'] as int?,
      status: json['status']?.toString() ?? 'unknown',
      senderEmail: json['sender_email'] as String?,
      subject: json['subject'] as String?,
      attachmentName: json['attachment_name'] as String?,
      supplierName: json['supplier_name'] as String?,
      supplierDocument: json['supplier_document'] as String?,
      recipientDocument: json['recipient_document'] as String?,
      invoiceKey: json['invoice_key'] as String?,
      invoiceNumber: json['invoice_number'] as String?,
      rejectionReason: json['rejection_reason'] as String?,
      receivedAt: DateTime.tryParse(json['received_at']?.toString() ?? ''),
    );
  }
}

File: backend/app/schemas/stock_entry.py
from datetime import date, datetime
from decimal import Decimal

from typing import Literal

from pydantic import BaseModel, Field


class StockEntryItemCreate(BaseModel):
    product_id: int | None = None
    description: str = Field(min_length=1, max_length=220)
    barcode: str | None = Field(default=None, max_length=80)
    tax_gtin: str | None = Field(default=None, max_length=80)
    supplier_product_code: str | None = Field(default=None, max_length=80)
    invoice_quantity: Decimal | None = Field(default=None, ge=0)
    invoice_unit: str | None = Field(default=None, max_length=20)
    tax_quantity: Decimal | None = Field(default=None, ge=0)
    tax_unit: str | None = Field(default=None, max_length=20)
    package_conversion_factor: Decimal | None = Field(default=None, gt=0)
    quantity: Decimal = Field(gt=0)
    received_quantity: Decimal | None = Field(default=None, ge=0)
    approved_quantity: Decimal | None = Field(default=None, ge=0)
    unit: str = Field(default="un", max_length=20)
    unit_cost: Decimal = Field(ge=0)
    total_cost: Decimal | None = Field(default=None, ge=0)
    ncm: str | None = Field(default=None, max_length=20)
    cfop: str | None = Field(default=None, max_length=10)
    origin: str | None = Field(default=None, max_length=2)
    cst: str | None = Field(default=None, max_length=10)
    csosn: str | None = Field(default=None, max_length=10)
    icms_rate: Decimal | None = Field(default=None, ge=0)
    pis_rate: Decimal | None = Field(default=None, ge=0)
    cofins_rate: Decimal | None = Field(default=None, ge=0)
    ipi_rate: Decimal | None = Field(default=None, ge=0)
    ibs_cbs_cst: str | None = Field(default=None, max_length=10)
    ibs_cbs_classification: str | None = Field(default=None, max_length=20)
    cbs_rate: Decimal | None = Field(default=None, ge=0)
    ibs_state_rate: Decimal | None = Field(default=None, ge=0)
    ibs_city_rate: Decimal | None = Field(default=None, ge=0)
    selective_tax_cst: str | None = Field(default=None, max_length=10)
    selective_tax_classification: str | None = Field(default=None, max_length=20)
    selective_tax_rate: Decimal | None = Field(default=None, ge=0)
    batch_number: str | None = Field(default=None, max_length=80)
    expiration_date: date | None = None
    manufacturing_date: date | None = None
    temperature_celsius: Decimal | None = None
    discrepancy_reason: str | None = Field(default=None, max_length=120)
    storage_location: str | None = Field(default=None, max_length=120)
    check_status: str = Field(default="accepted", max_length=30)
    check_notes: str | None = None


class StockEntryCreate(BaseModel):
    supplier_id: int | None = None
    supplier_name: str | None = Field(default=None, max_length=180)
    supplier_document: str | None = Field(default=None, max_length=30)
    source: str = Field(default="manual", max_length=30)
    invoice_key: str | None = Field(default=None, max_length=60)
    invoice_number: str | None = Field(default=None, max_length=30)
    invoice_series: str | None = Field(default=None, max_length=20)
    notes: str | None = None
    items: list[StockEntryItemCreate] = Field(min_length=1)


class StockEntryMobileReceiveItem(BaseModel):
    product_id: int | None = None
    description: str = Field(min_length=1, max_length=220)
    barcode: str | None = Field(default=None, max_length=80)
    quantity: Decimal = Field(gt=0)
    unit: str = Field(default="un", max_length=20)
    unit_cost: Decimal = Field(ge=0)
    sale_price: Decimal | None = Field(default=None, ge=0)
    ncm: str | None = Field(default=None, max_length=20)
    cfop: str | None = Field(default=None, max_length=10)
    origin: str | None = Field(default=None, max_length=2)
    cst: str | None = Field(default=None, max_length=10)
    csosn: str | None = Field(default=None, max_length=10)
    icms_rate: Decimal | None = Field(default=None, ge=0)
    pis_rate: Decimal | None = Field(default=None, ge=0)
    cofins_rate: Decimal | None = Field(default=None, ge=0)
    ipi_rate: Decimal | None = Field(default=None, ge=0)
    ibs_cbs_cst: str | None = Field(default=None, max_length=10)
    ibs_cbs_classification: str | None = Field(default=None, max_length=20)
    cbs_rate: Decimal | None = Field(default=None, ge=0)
    ibs_state_rate: Decimal | None = Field(default=None, ge=0)
    ibs_city_rate: Decimal | None = Field(default=None, ge=0)
    selective_tax_cst: str | None = Field(default=None, max_length=10)
    selective_tax_classification: str | None = Field(default=None, max_length=20)
    selective_tax_rate: Decimal | None = Field(default=None, ge=0)
    batch_number: str | None = Field(default=None, max_length=80)
    expiration_date: date | None = None
    manufacturing_date: date | None = None
    temperature_celsius: Decimal | None = None
    discrepancy_reason: str | None = Field(default=None, max_length=120)
    storage_location: str | None = Field(default=None, max_length=120)
    check_status: Literal["accepted", "return"] = "accepted"
    check_notes: str | None = None


class StockEntryItemMatchRequest(BaseModel):
    product_id: int = Field(gt=0)
    remember_supplier_link: bool = True
    supplier_product_code: str | None = Field(default=None, max_length=80)
    commercial_gtin: str | None = Field(default=None, max_length=80)
    tax_gtin: str | None = Field(default=None, max_length=80)
    supplier_description: str | None = Field(default=None, max_length=220)
    supplier_unit: str | None = Field(default=None, max_length=20)
    stock_unit: str | None = Field(default=None, max_length=20)
    conversion_factor: Decimal | None = Field(default=None, gt=0)


class StockEntryProductSuggestion(BaseModel):
    product_id: int
    name: str
    internal_code: str | None = None
    barcode: str | None = None
    purchase_package_barcode: str | None = None
    unit: str
    match_method: str
    warning: str | None = None


class StockEntryItemLotCreate(BaseModel):
    lot_number: str | None = Field(default=None, max_length=80)
    manufacturing_date: date | None = None
    expiration_date: date | None = None
    quantity: Decimal = Field(gt=0)
    temperature_celsius: Decimal | None = None
    notes: str | None = None


class StockEntryItemLotRead(BaseModel):
    id: int
    stock_entry_item_id: int
    lot_number: str | None
    manufacturing_date: date | None
    expiration_date: date | None
    quantity: Decimal
    temperature_celsius: Decimal | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class StockEntryItemRead(BaseModel):
    id: int
    product_id: int | None
    description: str
    barcode: str | None
    tax_gtin: str | None = None
    supplier_product_code: str | None = None
    invoice_quantity: Decimal | None = None
    invoice_unit: str | None = None
    tax_quantity: Decimal | None = None
    tax_unit: str | None = None
    package_conversion_factor: Decimal | None = None
    quantity: Decimal
    received_quantity: Decimal | None
    unit: str
    unit_cost: Decimal
    total_cost: Decimal
    ncm: str | None
    cfop: str | None
    origin: str | None
    cst: str | None
    csosn: str | None
    icms_rate: Decimal | None
    pis_rate: Decimal | None
    cofins_rate: Decimal | None
    ipi_rate: Decimal | None
    ibs_cbs_cst: str | None
    ibs_cbs_classification: str | None
    cbs_rate: Decimal | None
    ibs_state_rate: Decimal | None
    ibs_city_rate: Decimal | None
    selective_tax_cst: str | None
    selective_tax_classification: str | None
    selective_tax_rate: Decimal | None
    batch_number: str | None
    expiration_date: date | None
    manufacturing_date: date | None = None
    temperature_celsius: Decimal | None = None
    discrepancy_reason: str | None = None
    storage_location: str | None = None
    check_status: str
    check_notes: str | None
    lots: list["StockEntryItemLotRead"] = []
    created_at: datetime

    model_config = {"from_attributes": True}


class StockEntryRead(BaseModel):
    id: int
    supplier_id: int | None
    user_id: int | None
    source: str
    status: str
    invoice_key: str | None
    invoice_number: str | None
    invoice_series: str | None
    supplier_name: str | None
    supplier_document: str | None
    total_amount: Decimal
    notes: str | None
    confirmed_at: datetime | None
    reversed_at: datetime | None = None
    reversal_reason: str | None = None
    created_at: datetime
    items: list[StockEntryItemRead] = []

    model_config = {"from_attributes": True}


class StockDevolutionDraftItemRead(BaseModel):
    stock_entry_item_id: int
    product_id: int
    description: str
    quantity: Decimal
    unit: str
    unit_price: Decimal
    total_price: Decimal
    reason: str


class StockDevolutionDraftRead(BaseModel):
    fiscal_document_id: int
    stock_entry_id: int
    mode: str
    status: str
    supplier_name: str | None
    supplier_document: str | None
    original_invoice_number: str | None
    original_invoice_key: str | None
    total_amount: Decimal
    items: list[StockDevolutionDraftItemRead]


class StockReceiptPage(BaseModel):
    items: list[StockEntryRead]
    total: int
    page: int
    page_size: int


class StockReceiptSummary(BaseModel):
    all: int
    receiving: int
    pending_products: int
    divergences: int
    confirmed: int


class ReceivingSessionCreate(BaseModel):
    device_id: str | None = Field(default=None, max_length=120)


class StockEntryReversalRequest(BaseModel):
    reason: str = Field(min_length=5, max_length=500)


class ReceivingScanCreate(BaseModel):
    idempotency_key: str = Field(min_length=8, max_length=120)
    stock_entry_item_id: int | None = Field(default=None, gt=0)
    barcode: str | None = Field(default=None, max_length=80)
    quantity: Decimal = Field(gt=0)
    unit: str = Field(default="un", max_length=20)
    batch_number: str | None = Field(default=None, max_length=80)
    expiration_date: date | None = None
    condition: str = Field(default="ok", max_length=30)
    notes: str | None = None


class ReceivingSessionRead(BaseModel):
    id: int
    stock_entry_id: int
    user_id: int | None
    device_id: str | None
    status: str
    started_at: datetime
    finished_at: datetime | None
    scan_count: int = 0
    total_quantity: Decimal = Decimal("0")

    model_config = {"from_attributes": True}


class NfeXmlPreviewItem(BaseModel):
    product_id: int | None = None
    product_name: str | None = None
    description: str
    barcode: str | None = None
    tax_gtin: str | None = None
    supplier_product_code: str | None = None
    matching_status: str = "pending"
    matching_method: str | None = None
    match_suggestions: list[int] = []
    quantity: Decimal
    unit: str
    tax_quantity: Decimal | None = None
    tax_unit: str | None = None
    unit_cost: Decimal
    total_cost: Decimal
    ncm: str | None = None
    cfop: str | None = None
    origin: str | None = None
    cst: str | None = None
    csosn: str | None = None
    icms_rate: Decimal | None = None
    pis_rate: Decimal | None = None
    cofins_rate: Decimal | None = None
    ipi_rate: Decimal | None = None
    ibs_cbs_cst: str | None = None
    ibs_cbs_classification: str | None = None
    cbs_rate: Decimal | None = None
    ibs_state_rate: Decimal | None = None
    ibs_city_rate: Decimal | None = None
    selective_tax_cst: str | None = None
    selective_tax_classification: str | None = None
    selective_tax_rate: Decimal | None = None
    batch_number: str | None = None
    expiration_date: date | None = None


class NfeXmlPreview(BaseModel):
    supplier_id: int | None = None
    supplier_name: str | None = None
    supplier_document: str | None = None
    invoice_key: str | None = None
    invoice_number: str | None = None
    invoice_series: str | None = None
    items: list[NfeXmlPreviewItem]


class NfeXmlImportRequest(BaseModel):
    xml_content: str = Field(min_length=20)


class NfeKeyDownloadRequest(BaseModel):
    access_key: str = Field(min_length=44, max_length=60)
